File: navigation/safe-args-generator/src/tests/test-data/expected/nav_writer_test/MainFragmentArgs.java
/*
 * Copyright 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package a.b;

import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import java.io.Serializable;
import java.lang.IllegalArgumentException;
import java.lang.String;
import java.nio.file.AccessMode;

public class MainFragmentArgs {
    private Bundle bundle = new Bundle();

    private MainFragmentArgs() {
    }

    private MainFragmentArgs(Bundle bundle) {
        this.bundle.putAll(bundle);
    }

    @NonNull
    public static MainFragmentArgs fromBundle(Bundle bundle) {
        MainFragmentArgs result = new MainFragmentArgs();
        bundle.setClassLoader(MainFragmentArgs.class.getClassLoader());
        if (bundle.containsKey("main")) {
            String main;
            main = bundle.getString("main");
            if (main == null) {
                throw new IllegalArgumentException("Argument \"main\" is marked as non-null but was passed a null value.");
            }
            result.bundle.putString("main", main);
        } else {
            throw new IllegalArgumentException("Required argument \"main\" is missing and does not have an android:defaultValue");
        }
        if (bundle.containsKey("optional")) {
            int optional;
            optional = bundle.getInt("optional");
            result.bundle.putInt("optional", optional);
        }
        if (bundle.containsKey("reference")) {
            int reference;
            reference = bundle.getInt("reference");
            result.bundle.putInt("reference", reference);
        }
        if (bundle.containsKey("floatArg")) {
            float floatArg;
            floatArg = bundle.getFloat("floatArg");
            result.bundle.putFloat("floatArg", floatArg);
        }
        if (bundle.containsKey("floatArrayArg")) {
            float[] floatArrayArg;
            floatArrayArg = bundle.getFloatArray("floatArrayArg");
            if (floatArrayArg == null) {
                throw new IllegalArgumentException("Argument \"floatArrayArg\" is marked as non-null but was passed a null value.");
            }
            result.bundle.putFloatArray("floatArrayArg", floatArrayArg);
        } else {
            throw new IllegalArgumentException("Required argument \"floatArrayArg\" is missing and does not have an android:defaultValue");
        }
        if (bundle.containsKey("objectArrayArg")) {
            ActivityInfo[] objectArrayArg;
            objectArrayArg = (ActivityInfo[]) bundle.getParcelableArray("objectArrayArg");
            if (objectArrayArg == null) {
                throw new IllegalArgumentException("Argument \"objectArrayArg\" is marked as non-null but was passed a null value.");
            }
            result.bundle.putParcelableArray("objectArrayArg", objectArrayArg);
        } else {
            throw new IllegalArgumentException("Required argument \"objectArrayArg\" is missing and does not have an android:defaultValue");
        }
        if (bundle.containsKey("boolArg")) {
            boolean boolArg;
            boolArg = bundle.getBoolean("boolArg");
            result.bundle.putBoolean("boolArg", boolArg);
        }
        if (bundle.containsKey("optionalParcelable")) {
            ActivityInfo optionalParcelable;
            if (Parcelable.class.isAssignableFrom(ActivityInfo.class) || Serializable.class.isAssignableFrom(ActivityInfo.class)) {
                optionalParcelable = (ActivityInfo) bundle.get("optionalParcelable");
            } else {
                throw new UnsupportedOperationException(ActivityInfo.class.getName() + " must implement Parcelable or Serializable or must be an Enum.");
            }
            if (Parcelable.class.isAssignableFrom(ActivityInfo.class) || optionalParcelable == null) {
                result.bundle.putParcelable("optionalParcelable", Parcelable.class.cast(optionalParcelable));
            } else if (Serializable.class.isAssignableFrom(ActivityInfo.class)) {
                result.bundle.putSerializable("optionalParcelable", Serializable.class.cast(optionalParcelable));
            } else {
                throw new UnsupportedOperationException(ActivityInfo.class.getName() + " must implement Parcelable or Serializable or must be an Enum.");
            }
        }
        if (bundle.containsKey("enumArg")) {
            AccessMode enumArg;
            if (Parcelable.class.isAssignableFrom(AccessMode.class) || Serializable.class.isAssignableFrom(AccessMode.class)) {
                enumArg = (AccessMode) bundle.get("enumArg");
            } else {
                throw new UnsupportedOperationException(AccessMode.class.getName() + " must implement Parcelable or Serializable or must be an Enum.");
            }
            if (enumArg == null) {
                throw new IllegalArgumentException("Argument \"enumArg\" is marked as non-null but was passed a null value.");
            }
            if (Parcelable.class.isAssignableFrom(AccessMode.class) || enumArg == null) {
                result.bundle.putParcelable("enumArg", Parcelable.class.cast(enumArg));
            } else if (Serializable.class.isAssignableFrom(AccessMode.class)) {
                result.bundle.putSerializable("enumArg", Serializable.class.cast(enumArg));
            } else {
                throw new UnsupportedOperationException(AccessMode.class.getName() + " must implement Parcelable or Serializable or must be an Enum.");
            }
        }
        return result;
    }

    @NonNull
    public String getMain() {
        String main;
        main = bundle.getString("main");
        return main;
    }

    public int getOptional() {
        int optional;
        optional = bundle.getInt("optional");
        return optional;
    }

    public int getReference() {
        int reference;
        reference = bundle.getInt("reference");
        return reference;
    }

    public float getFloatArg() {
        float floatArg;
        floatArg = bundle.getFloat("floatArg");
        return floatArg;
    }

    @NonNull
    public float[] getFloatArrayArg() {
        float[] floatArrayArg;
        floatArrayArg = bundle.getFloatArray("floatArrayArg");
        return floatArrayArg;
    }

    @NonNull
    public ActivityInfo[] getObjectArrayArg() {
        ActivityInfo[] objectArrayArg;
        objectArrayArg = (ActivityInfo[]) bundle.getParcelableArray("objectArrayArg");
        return objectArrayArg;
    }

    public boolean getBoolArg() {
        boolean boolArg;
        boolArg = bundle.getBoolean("boolArg");
        return boolArg;
    }

    @Nullable
    public ActivityInfo getOptionalParcelable() {
        ActivityInfo optionalParcelable;
        if (Parcelable.class.isAssignableFrom(ActivityInfo.class) || Serializable.class.isAssignableFrom(ActivityInfo.class)) {
            optionalParcelable = (ActivityInfo) bundle.get("optionalParcelable");
        } else {
            throw new UnsupportedOperationException(ActivityInfo.class.getName() + " must implement Parcelable or Serializable or must be an Enum.");
        }
        return optionalParcelable;
    }

    @NonNull
    public AccessMode getEnumArg() {
        AccessMode enumArg;
        if (Parcelable.class.isAssignableFrom(AccessMode.class) || Serializable.class.isAssignableFrom(AccessMode.class)) {
            enumArg = (AccessMode) bundle.get("enumArg");
        } else {
            throw new UnsupportedOperationException(AccessMode.class.getName() + " must implement Parcelable or Serializable or must be an Enum.");
        }
        return enumArg;
    }

    @NonNull
    public Bundle toBundle() {
        return bundle;
    }

    public static class Builder {
        private Bundle bundle = new Bundle();

        public Builder(MainFragmentArgs original) {
            this.bundle.putAll(original.bundle);
        }

        public Builder(@NonNull String main, @NonNull float[] floatArrayArg,
                @NonNull ActivityInfo[] objectArrayArg) {
            if (main == null) {
                throw new IllegalArgumentException("Argument \"main\" is marked as non-null but was passed a null value.");
            }
            bundle.putString("main", main);
            if (floatArrayArg == null) {
                throw new IllegalArgumentException("Argument \"floatArrayArg\" is marked as non-null but was passed a null value.");
            }
            bundle.putFloatArray("floatArrayArg", floatArrayArg);
            if (objectArrayArg == null) {
                throw new IllegalArgumentException("Argument \"objectArrayArg\" is marked as non-null but was passed a null value.");
            }
            bundle.putParcelableArray("objectArrayArg", objectArrayArg);
        }

        @NonNull
        public MainFragmentArgs build() {
            MainFragmentArgs result = new MainFragmentArgs(bundle);
            return result;
        }

        @NonNull
        public Builder setMain(@NonNull String main) {
            if (main == null) {
                throw new IllegalArgumentException("Argument \"main\" is marked as non-null but was passed a null value.");
            }
            this.bundle.putString("main", main);
            return this;
        }

        @NonNull
        public Builder setOptional(int optional) {
            this.bundle.putInt("optional", optional);
            return this;
        }

        @NonNull
        public Builder setReference(int reference) {
            this.bundle.putInt("reference", reference);
            return this;
        }

        @NonNull
        public Builder setFloatArg(float floatArg) {
            this.bundle.putFloat("floatArg", floatArg);
            return this;
        }

        @NonNull
        public Builder setFloatArrayArg(@NonNull float[] floatArrayArg) {
            if (floatArrayArg == null) {
                throw new IllegalArgumentException("Argument \"floatArrayArg\" is marked as non-null but was passed a null value.");
            }
            this.bundle.putFloatArray("floatArrayArg", floatArrayArg);
            return this;
        }

        @NonNull
        public Builder setObjectArrayArg(@NonNull ActivityInfo[] objectArrayArg) {
            if (objectArrayArg == null) {
                throw new IllegalArgumentException("Argument \"objectArrayArg\" is marked as non-null but was passed a null value.");
            }
            this.bundle.putParcelableArray("objectArrayArg", objectArrayArg);
            return this;
        }

        @NonNull
        public Builder setBoolArg(boolean boolArg) {
            this.bundle.putBoolean("boolArg", boolArg);
            return this;
        }

        @NonNull
        public Builder setOptionalParcelable(@Nullable ActivityInfo optionalParcelable) {
            if (Parcelable.class.isAssignableFrom(ActivityInfo.class) || optionalParcelable == null) {
                this.bundle.putParcelable("optionalParcelable", Parcelable.class.cast(optionalParcelable));
            } else if (Serializable.class.isAssignableFrom(ActivityInfo.class)) {
                this.bundle.putSerializable("optionalParcelable", Serializable.class.cast(optionalParcelable));
            } else {
                throw new UnsupportedOperationException(ActivityInfo.class.getName() + " must implement Parcelable or Serializable or must be an Enum.");
            }
            return this;
        }

        @NonNull
        public Builder setEnumArg(@NonNull AccessMode enumArg) {
            if (enumArg == null) {
                throw new IllegalArgumentException("Argument \"enumArg\" is marked as non-null but was passed a null value.");
            }
            if (Parcelable.class.isAssignableFrom(AccessMode.class) || enumArg == null) {
                this.bundle.putParcelable("enumArg", Parcelable.class.cast(enumArg));
            } else if (Serializable.class.isAssignableFrom(AccessMode.class)) {
                this.bundle.putSerializable("enumArg", Serializable.class.cast(enumArg));
            } else {
                throw new UnsupportedOperationException(AccessMode.class.getName() + " must implement Parcelable or Serializable or must be an Enum.");
            }
            return this;
        }

        @NonNull
        public String getMain() {
            String main;
            main = bundle.getString("main");
            return main;
        }

        public int getOptional() {
            int optional;
            optional = bundle.getInt("optional");
            return optional;
        }

        public int getReference() {
            int reference;
            reference = bundle.getInt("reference");
            return reference;
        }

        public float getFloatArg() {
            float floatArg;
            floatArg = bundle.getFloat("floatArg");
            return floatArg;
        }

        @NonNull
        public float[] getFloatArrayArg() {
            float[] floatArrayArg;
            floatArrayArg = bundle.getFloatArray("floatArrayArg");
            return floatArrayArg;
        }

        @NonNull
        public ActivityInfo[] getObjectArrayArg() {
            ActivityInfo[] objectArrayArg;
            objectArrayArg = (ActivityInfo[]) bundle.getParcelableArray("objectArrayArg");
            return objectArrayArg;
        }

        public boolean getBoolArg() {
            boolean boolArg;
            boolArg = bundle.getBoolean("boolArg");
            return boolArg;
        }

        @Nullable
        public ActivityInfo getOptionalParcelable() {
            ActivityInfo optionalParcelable;
            if (Parcelable.class.isAssignableFrom(ActivityInfo.class) || Serializable.class.isAssignableFrom(ActivityInfo.class)) {
                optionalParcelable = (ActivityInfo) bundle.get("optionalParcelable");
            } else {
                throw new UnsupportedOperationException(ActivityInfo.class.getName() + " must implement Parcelable or Serializable or must be an Enum.");
            }
            return optionalParcelable;
        }

        @NonNull
        public AccessMode getEnumArg() {
            AccessMode enumArg;
            if (Parcelable.class.isAssignableFrom(AccessMode.class) || Serializable.class.isAssignableFrom(AccessMode.class)) {
                enumArg = (AccessMode) bundle.get("enumArg");
            } else {
                throw new UnsupportedOperationException(AccessMode.class.getName() + " must implement Parcelable or Serializable or must be an Enum.");
            }
            return enumArg;
        }
    }
}

File: navigation/safe-args-generator/src/tests/test-data/expected/kotlin_nav_writer_test/MainFragmentArgs.kt
package a.b

import android.content.pm.ActivityInfo
import android.os.Bundle
import android.os.Parcelable
import androidx.navigation.NavArgs
import java.io.Serializable
import java.lang.IllegalArgumentException
import java.lang.UnsupportedOperationException
import java.nio.file.AccessMode
import kotlin.Array
import kotlin.Boolean
import kotlin.Float
import kotlin.FloatArray
import kotlin.Int
import kotlin.String
import kotlin.Suppress

data class MainFragmentArgs(
    val main: String,
    val optional: Int = -1,
    val reference: Int = a.b.R.drawable.background,
    val floatArg: Float = 1F,
    val floatArrayArg: FloatArray,
    val objectArrayArg: Array<ActivityInfo>,
    val boolArg: Boolean = true,
    val optionalParcelable: ActivityInfo? = null,
    val enumArg: AccessMode = AccessMode.READ
) : NavArgs {
    @Suppress("UNCHECKED_CAST", "USELESS_CAST")
    constructor(bundle: Bundle) : this(bundle.let {
        it.setClassLoader(MainFragmentArgs::class.java.classLoader)
        val __main : String?
        if (it.containsKey("main")) {
            __main = it.getString("main")
            if (__main == null) {
                throw IllegalArgumentException("Argument \"main\" is marked as non-null but was passed a null value.")
            }
        } else {
            throw IllegalArgumentException("Required argument \"main\" is missing and does not have an android:defaultValue")
        }
        return@let __main as String
    }
    , bundle.let {
        it.setClassLoader(MainFragmentArgs::class.java.classLoader)
        val __optional : Int
        if (it.containsKey("optional")) {
            __optional = it.getInt("optional")
        } else {
            __optional = -1
        }
        return@let __optional as Int
    }
    , bundle.let {
        it.setClassLoader(MainFragmentArgs::class.java.classLoader)
        val __reference : Int
        if (it.containsKey("reference")) {
            __reference = it.getInt("reference")
        } else {
            __reference = a.b.R.drawable.background
        }
        return@let __reference as Int
    }
    , bundle.let {
        it.setClassLoader(MainFragmentArgs::class.java.classLoader)
        val __floatArg : Float
        if (it.containsKey("floatArg")) {
            __floatArg = it.getFloat("floatArg")
        } else {
            __floatArg = 1F
        }
        return@let __floatArg as Float
    }
    , bundle.let {
        it.setClassLoader(MainFragmentArgs::class.java.classLoader)
        val __floatArrayArg : FloatArray?
        if (it.containsKey("floatArrayArg")) {
            __floatArrayArg = it.getFloatArray("floatArrayArg")
            if (__floatArrayArg == null) {
                throw IllegalArgumentException("Argument \"floatArrayArg\" is marked as non-null but was passed a null value.")
            }
        } else {
            throw IllegalArgumentException("Required argument \"floatArrayArg\" is missing and does not have an android:defaultValue")
        }
        return@let __floatArrayArg as FloatArray
    }
    , bundle.let {
        it.setClassLoader(MainFragmentArgs::class.java.classLoader)
        val __objectArrayArg : Array<ActivityInfo>?
        if (it.containsKey("objectArrayArg")) {
            __objectArrayArg = it.getParcelableArray("objectArrayArg") as Array<ActivityInfo>?
            if (__objectArrayArg == null) {
                throw IllegalArgumentException("Argument \"objectArrayArg\" is marked as non-null but was passed a null value.")
            }
        } else {
            throw IllegalArgumentException("Required argument \"objectArrayArg\" is missing and does not have an android:defaultValue")
        }
        return@let __objectArrayArg as Array<ActivityInfo>
    }
    , bundle.let {
        it.setClassLoader(MainFragmentArgs::class.java.classLoader)
        val __boolArg : Boolean
        if (it.containsKey("boolArg")) {
            __boolArg = it.getBoolean("boolArg")
        } else {
            __boolArg = true
        }
        return@let __boolArg as Boolean
    }
    , bundle.let {
        it.setClassLoader(MainFragmentArgs::class.java.classLoader)
        val __optionalParcelable : ActivityInfo?
        if (it.containsKey("optionalParcelable")) {
            if (Parcelable::class.java.isAssignableFrom(ActivityInfo::class.java) ||
                    Serializable::class.java.isAssignableFrom(ActivityInfo::class.java)) {
                __optionalParcelable = it.get("optionalParcelable") as ActivityInfo?
            } else {
                throw UnsupportedOperationException(ActivityInfo::class.java.name +
                        " must implement Parcelable or Serializable or must be an Enum.")
            }
        } else {
            __optionalParcelable = null
        }
        return@let __optionalParcelable as ActivityInfo?
    }
    , bundle.let {
        it.setClassLoader(MainFragmentArgs::class.java.classLoader)
        val __enumArg : AccessMode?
        if (it.containsKey("enumArg")) {
            if (Parcelable::class.java.isAssignableFrom(AccessMode::class.java) ||
                    Serializable::class.java.isAssignableFrom(AccessMode::class.java)) {
                __enumArg = it.get("enumArg") as AccessMode?
            } else {
                throw UnsupportedOperationException(AccessMode::class.java.name +
                        " must implement Parcelable or Serializable or must be an Enum.")
            }
            if (__enumArg == null) {
                throw IllegalArgumentException("Argument \"enumArg\" is marked as non-null but was passed a null value.")
            }
        } else {
            __enumArg = AccessMode.READ
        }
        return@let __enumArg as AccessMode
    }
    )

    @Suppress("CAST_NEVER_SUCCEEDS")
    fun toBundle(): Bundle {
        val __result = Bundle()
        __result.putString("main", main)
        __result.putInt("optional", optional)
        __result.putInt("reference", reference)
        __result.putFloat("floatArg", floatArg)
        __result.putFloatArray("floatArrayArg", floatArrayArg)
        __result.putParcelableArray("objectArrayArg", objectArrayArg)
        __result.putBoolean("boolArg", boolArg)
        if (Parcelable::class.java.isAssignableFrom(ActivityInfo::class.java)) {
            __result.putParcelable("optionalParcelable", optionalParcelable as Parcelable?)
        } else if (Serializable::class.java.isAssignableFrom(ActivityInfo::class.java)) {
            __result.putSerializable("optionalParcelable", optionalParcelable as Serializable?)
        }
        if (Parcelable::class.java.isAssignableFrom(AccessMode::class.java)) {
            __result.putParcelable("enumArg", enumArg as Parcelable)
        } else if (Serializable::class.java.isAssignableFrom(AccessMode::class.java)) {
            __result.putSerializable("enumArg", enumArg as Serializable)
        }
        return __result
    }
}
